src/ctapipe/image/modifications.py

```diff
@@ -100,6 +111,195 @@ def _smear_psf_randomly(
+        # n_pixels, n_samples]
+        source = EventSource(input_url=self.nsb_file, skip_calibration_events=False)
+        nsb_database = defaultdict(list)  # [nevents, ngains, npixels, nsamples]
+        for event in source:
+            if not event_type_filter(event):
+                continue
+            for tel_id in event.trigger.tels_with_trigger:
+                nsb_database[tel_id].append(event.r1.tel[tel_id].waveform)
+        nsb_database = {
+            tel_id: np.stack(waveforms) for tel_id, waveforms in nsb_database.items()
+        }
+
+        # Check that we have enough NSB-only events for all telescopes. We
+        # require that the number of NSB events for any telescope is at
+        # least two times the number of waveforms (=nsb_level) that we will add
+        # up. This is to avoid excessive correlation among the waveforms.
+        stats_ok = True
+        for tel_id in nsb_database:
+            nevents = nsb_database[tel_id].shape[0]
+            if nevents >= 2 * self.nsb_level:
+                continue
+            self.log.error(
+                f"Not enough NSB events available for tel_"
+                f"id {tel_id}. "
+                f"For nsb_level = {self.nsb_level}, at least "
+                f"{2*self.nsb_level} events are needed ({nevents} "
+                f"were found)."
+            )
+            stats_ok = False
+
+        if not stats_ok:
+            raise ValueError("Please use an input nsb_file with more events!")
+
+        # Now shift the waveforms so that they have mean=0 and do not introduce
+        # any bias (just fluctuations). For each telescope and gain we average
+        # all pixels
+        for tel_id in nsb_database:
+            hg_mean = np.mean(nsb_database[tel_id][:, 0, :, :])  # HG
+            lg_mean = np.mean(nsb_database[tel_id][:, 1, :, :])  # LG
+            nsb_database[tel_id][:, 0, :, :] -= hg_mean
+            nsb_database[tel_id][:, 1, :, :] -= lg_mean
+
+        # Now add up waveforms selected at random to obtain different
+        # realizations of the total noise that will be added
+        for tel_id in nsb_database:
+            newshape = np.array(nsb_database[tel_id].shape)
+            newshape[0] = self.n_noise_realizations
+            self.total_noise[tel_id] = np.zeros(shape=newshape)
+            # [n_noise_realizations, ngains, npixels, nsamples]
+
+            nevents = nsb_database[tel_id].shape[0]
+            for jj in range(self.n_noise_realizations):
+                # We take a random combination of a number self.nsb_level of
+                # instances of the noise events. The combination of events is
+                # common to the whole camera (making it different for each
+                # pixel might be better, but it is very slow - and I failed
+                # to do it using numba due to some numba limitations)
+                shuffled_event_indices = self.rng.permutation(nevents)[: self.nsb_level]
```

Comment:
What about something like:
```python
suffled_event_indices = rng.choice(n_events, size=(self.n_noise_realizations, self.nsb_level))
```
That gives you a numpy array of the indices for each permutation. That is much faster and could be used to get a different shuffling per pixel if you wanted.   Since this is not run frequently, however, speed is not a giant concern.

Reply:
What is `n_noise_realizations`? You pre-compute the full noise waveforms before-hand?

Reply:
For 100 noise events out of 10000 events, this takes 2 seconds. Vectorizing the sum takes it down to 1.4:

```
for i in range(n_noise):
    for pixel in range(n_pixels):
        chosen = rng.choice(n_events, size=nsb_level, replace=False)
        noise[i, :, pixel] += waveforms[chosen, :, pixel].sum(axis=0)
```

Reply:
Ah, ok, indeed it is the same, I thought you were producing as many noise realizations as the number of "elemental noise events".
For the record: n_events is the number of "elemental" noise events we have, n_noise_realizations is the number of combinations of them we want to make, nsb_level the number of elemental waveforms we combine in one of the total_noise waveforms.

Reply:
Will try it.